source/Examples/SmartCache/README.md

```diff
@@ -0,0 +1,115 @@
+
+TODO: Some description of the example contained in this directory.
+
+This `SmartCache` can be used to demonstrate a full end-to-end MLOS integrated microbenchmark for a "smart" component (in this case a cache).
+
+## Overview
+
+To do that, we run the (C++) `SmartCache` executable to communicate with the (C#) [`Mlos.Agent.Server`](../../Mlos.Agent.Server/) over a shared memory channel provided by the [`Mlos.Core`](../../Mlos.Core/) library.
+
+The `Mlos.Agent.Server` is essentially a small wrapper around several other communication channels to allow different components to connect for component experimentation convenience.
+
+It provides
+
+1. Shared memory communication channels via the [`Mlos.Agent`](../../Mlos.Agent) and [`Mlos.NetCore`](../../Mlos.NetCore) libraries.
+
+2. A [`Mlos.Agent.GrpcServer`](../../Mlos.Agent.GrpcClient/) GRPC channel to allow driving the experimentation process from a Jupyter notebook.
+
+3. A GRPC client to connect to the (Python) [`mlos.Grpc.OptimizerMicroserviceServer`](../../Mlos.Python/mlos/Grpc/OptimizerMicroserviceServer.py) to store and track those experiments.
+
+TODO: Diagrams
+
+## Building
+
+To build and run the necessary components for this example
+
+1. [Build the Docker image](../../../documentation/01-Prerequisites.md#build-the-docker-image) using the [`Dockerfile`](../../../Dockerfile) at the root of the repository.
+
+2. [Run the Docker image](../../../documentation/02-Build.md#create-a-new-container-instance) you just built.
+
+3. Inside the container, [build the compiled software](../../../documentation/02-Build.md#cli-make) with `make`:
+
+    ```sh
+    make dotnet-build cmake-build
+    ```
+
+    > This will build everything using a default `CONFIGURATION=Release`.
+    >
+    > To just build `SmartCache` and `Mlos.Agent.Server`, execute the following: \
+    > `make -C source/Examples/SmartCache && make -C source/Mlos.Agent.Server`
+
+4. For a `Release` build, the relevant output will be at:
+
+    - `Mlos.Agent.Server`:
+
+        `out/dotnet/source/Mlos.Agent.Server/obj/AnyCPU/Mlos.Agent.Server.dll`
+
+    - `SmartCache`:
+
+        `out/cmake/Release/source/Examples/SmartCache/SmartCache`
+
+    - `SmartCache.SettingsRegistry`
+
+        `out/dotnet/source/Examples/SmartCache/SmartCache.SettingsRegistry/obj/AnyCPU/SmartCache.SettingsRegistry.dll`
+
+## Executing
+
+> Note: these commands are given relative to the root of the MLOS repo.
+>
+> To move there, you can execute the following within the repository:
+>
+> `cd $(git rev-parse --show-toplevel)`
+
+`SmartCache` can be invoked separate, or by the `Mlos.Agent.Server` itself.
+
+Once started, `SmartCache` will attempt to register its component specific set of shared memory messages with the `Mlos.Agent` in the `Mlos.Agent.Server` using some `Mlos.Core` component registration messages.  That message includes the name of the `SettingsRegistry` assembly (`.dll`) corresponding to that component's settings/messages.
```

Comment:
I'm not sure how to fix this, but right now I got confused by the order of things. You're giving a high level overview and then you're giving the actual commands, which is good. But I didn't understand that on the first read and so I was like "you expected me to start SmartCache already but I didn't know how to".
In particular the "Executing" part starts with a very low-level instruction before going to a highlevel overview.

Maybe skipping the high-level overview in this place is easiest and do it inline with the commands? But that's just a didactic thing, we can polish this later.

Reply:
I tried to reorganize this with a more simplified "command first, then explain" format.  Please have another look and let me know what you think.